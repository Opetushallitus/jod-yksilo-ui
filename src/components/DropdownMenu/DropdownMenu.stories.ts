import type { Meta, StoryObj } from '@storybook/react';

import { DropdownMenu } from './DropdownMenu';

const meta = {
  title: 'Primitives/DropdownMenu',
  component: DropdownMenu,
  tags: ['autodocs'],
} satisfies Meta<typeof DropdownMenu>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: 'Choose language',
    options: [
      { label: 'Suomi', value: 'fi' },
      { label: 'Svenska', value: 'sv' },
      { label: 'English', value: 'en' },
    ],
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
    label: 'Choose language',
    options: [
      { label: 'Suomi', value: 'fi' },
      { label: 'Svenska', value: 'sv' },
      { label: 'English', value: 'en' },
    ],
  },
};
