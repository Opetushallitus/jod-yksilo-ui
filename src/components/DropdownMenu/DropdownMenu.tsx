import { useId } from 'react';
import classNames from 'classnames';

export interface DropdownMenuOptionsData<T extends string = string> {
  value: T;
  label: string;
}

interface DropdownMenuProps<T extends DropdownMenuOptionsData, U extends string = string> {
  /** Label for the component */
  label: string;
  /** Hide label. Still available for screenreaders */
  hideLabel?: boolean;
  /** Options for component */
  options: T[];
  /** Default value to be selected initially */
  defaultValue?: U;
  /** Controlled mode */
  selected?: U;
  /** Callback on selection change */
  onChange?: (value: U) => void;
  /** Component is disabled for user interaction */
  disabled?: boolean;
}

export const DropdownMenu = <
  U extends string = string,
  T extends DropdownMenuOptionsData<string> = DropdownMenuOptionsData<string>,
>({
  label,
  hideLabel = false,
  options,
  defaultValue,
  onChange: propOnChange,
  disabled = false,
}: DropdownMenuProps<T, U>) => {
  const labelId = useId();
  return (
    <div className="flex flex-row">
      <label
        htmlFor={labelId}
        className={classNames('mr-2 self-center font-bold text-jod-black', {
          'sr-only': hideLabel,
        })}
      >
        {label}
      </label>
      <select
        disabled={disabled}
        id={labelId}
        className={classNames(
          'min-w-[120px] justify-self-end rounded-lg border border-jod-dark bg-jod-white p-2 hover:bg-purple-100 focus:outline-none focus:ring focus:ring-purple-500',
          {
            'disabled:border-gray-500 disabled:bg-gray-200 disabled:text-gray-500 disabled:hover:bg-gray-200': disabled,
          },
        )}
        onChange={(event: React.ChangeEvent<HTMLSelectElement>) => {
          if (propOnChange) {
            propOnChange(event.target.value as U);
          }
        }}
        defaultValue={defaultValue}
      >
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  );
};
