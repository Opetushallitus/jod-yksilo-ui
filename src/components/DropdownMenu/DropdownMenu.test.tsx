import { afterEach, describe, expect, it, vi } from 'vitest';
import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DropdownMenu } from './DropdownMenu';

afterEach(() => {
  cleanup();
});

describe('Snapshot testing', () => {
  it('should render with defaults', () => {
    const { container } = render(<DropdownMenu label="" options={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not show label', () => {
    const { container } = render(<DropdownMenu hideLabel={true} label="" options={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render as disabled', () => {
    const { container } = render(<DropdownMenu disabled={true} label="" options={[]} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultValue', () => {
  const options = [
    { value: 'fi', label: 'Suomi' },
    { value: 'sv', label: 'Svenska' },
    { value: 'en', label: 'English' },
  ];

  it('should select first by default', () => {
    render(<DropdownMenu label="" options={options} />);
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'Suomi' }) as HTMLOptionElement).selected).toBe(true);
  });

  it('should select the wanted one', () => {
    render(<DropdownMenu label="" options={options} defaultValue="sv" />);
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'Svenska' }) as HTMLOptionElement).selected).toBe(true);
    //
  });

  it('should still allow user to another', async () => {
    render(<DropdownMenu label="" options={options} defaultValue="fi" />);
    const user = userEvent.setup();
    await user.selectOptions(screen.getByRole('combobox'), 'English');

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    expect((screen.getByRole('option', { name: 'English' }) as HTMLOptionElement).selected).toBe(true);
  });
});

describe('onChange', () => {
  const options = [
    { value: 'fi', label: 'Suomi' },
    { value: 'sv', label: 'Svenska' },
    { value: 'en', label: 'English' },
  ];

  it('is called on user selection', async () => {
    const mockCallback = vi.fn();
    const user = userEvent.setup();

    render(<DropdownMenu label="" options={options} onChange={mockCallback} />);

    await user.selectOptions(screen.getByRole('combobox'), 'English');
    expect(mockCallback).toHaveBeenCalledTimes(1);
  });

  it('is not called when disabled', async () => {
    const mockCallback = vi.fn();
    const user = userEvent.setup();

    render(<DropdownMenu label="" options={options} onChange={mockCallback} disabled={true} />);

    await user.selectOptions(screen.getByRole('combobox'), 'English');
    expect(mockCallback).toHaveBeenCalledTimes(0);
  });
});

describe('label', () => {
  it('should render with label', () => {
    render(<DropdownMenu label="Label" options={[]} />);
    expect(screen.getByLabelText('Label')).not.toBeNull();
  });
});

describe('hideLabel', () => {
  it('should add correct CSS class to label', () => {
    render(<DropdownMenu label="Non-visible label" options={[]} hideLabel={true} />);
    expect(screen.getByText('Non-visible label').classList).toContain('sr-only');
  });
});
