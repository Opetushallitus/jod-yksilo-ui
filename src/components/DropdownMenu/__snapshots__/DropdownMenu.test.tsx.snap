// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > should not show label 1`] = `
<div
  class="flex flex-row"
>
  <label
    class="mr-2 self-center font-bold text-jod-black sr-only"
    for=":r1:"
  />
  <select
    class="min-w-[120px] justify-self-end rounded-lg border border-jod-dark bg-jod-white p-2 hover:bg-purple-100 focus:outline-none focus:ring focus:ring-purple-500"
    id=":r1:"
  />
</div>
`;

exports[`Snapshot testing > should render as disabled 1`] = `
<div
  class="flex flex-row"
>
  <label
    class="mr-2 self-center font-bold text-jod-black"
    for=":r2:"
  />
  <select
    class="min-w-[120px] justify-self-end rounded-lg border border-jod-dark bg-jod-white p-2 hover:bg-purple-100 focus:outline-none focus:ring focus:ring-purple-500 disabled:border-gray-500 disabled:bg-gray-200 disabled:text-gray-500 disabled:hover:bg-gray-200"
    disabled=""
    id=":r2:"
  />
</div>
`;

exports[`Snapshot testing > should render with defaults 1`] = `
<div
  class="flex flex-row"
>
  <label
    class="mr-2 self-center font-bold text-jod-black"
    for=":r0:"
  />
  <select
    class="min-w-[120px] justify-self-end rounded-lg border border-jod-dark bg-jod-white p-2 hover:bg-purple-100 focus:outline-none focus:ring focus:ring-purple-500"
    id=":r0:"
  />
</div>
`;
